cli/ballerina-launcher/src/main/java/org/ballerinalang/launcher/Main.java

```diff
@@ -80,77 +75,73 @@ public static void main(String... args) {
             defaultCmd.setParentCmdParser(cmdParser);
 
             // Run command
             RunCmd runCmd = new RunCmd();
-            JCommander jcRunCmd = addSubCommand(cmdParser, BallerinaCliCommands.RUN, runCmd);
+            CommandLine jcRunCmd = addSubCommand(cmdParser, BallerinaCliCommands.RUN, runCmd);
             runCmd.setParentCmdParser(cmdParser);
             runCmd.setSelfCmdParser(jcRunCmd);
 
             HelpCmd helpCmd = new HelpCmd();
-            cmdParser.addCommand(BallerinaCliCommands.HELP, helpCmd);
+            cmdParser.addSubcommand(BallerinaCliCommands.HELP, helpCmd);
             helpCmd.setParentCmdParser(cmdParser);
 
             // loading additional commands via SPI
             ServiceLoader<BLauncherCmd> bCmds = ServiceLoader.load(BLauncherCmd.class);
             for (BLauncherCmd bCmd : bCmds) {
-                cmdParser.addCommand(bCmd.getName(), bCmd);
+                cmdParser.addSubcommand(bCmd.getName(), bCmd);
                 bCmd.setParentCmdParser(cmdParser);
             }
 
             // Build Version Command
             VersionCmd versionCmd = new VersionCmd();
-            cmdParser.addCommand(BallerinaCliCommands.VERSION, versionCmd);
+            cmdParser.addSubcommand(BallerinaCliCommands.VERSION, versionCmd);
             versionCmd.setParentCmdParser(cmdParser);
 
             EncryptCmd encryptCmd = new EncryptCmd();
-            cmdParser.addCommand(BallerinaCliCommands.ENCRYPT, encryptCmd);
+            cmdParser.addSubcommand(BallerinaCliCommands.ENCRYPT, encryptCmd);
             encryptCmd.setParentCmdParser(cmdParser);
 
-            cmdParser.setProgramName("ballerina");
-            cmdParser.parse(args);
-            String parsedCmdName = cmdParser.getParsedCommand();
+            cmdParser.setCommandName("ballerina");
 
-            // User has not specified a command. Therefore returning the main command
-            // which simply prints usage information.
-            if (parsedCmdName == null) {
-                return Optional.of(defaultCmd);
-            }
+            cmdParser.setStopAtPositional(true);
 
-            Map<String, JCommander> commanderMap = cmdParser.getCommands();
-            return Optional.of((BLauncherCmd) commanderMap.get(parsedCmdName).getObjects().get(0));
+            List<CommandLine> parsedCommands = cmdParser.parse(args);
 
-        } catch (MissingCommandException e) {
-            String errorMsg = "unknown command '" + e.getUnknownCommand() + "'";
-            throw LauncherUtils.createUsageException(errorMsg);
+            if (parsedCommands.size() < 1) {
+                return Optional.of(defaultCmd);
+            }
 
-        } catch (ParameterException e) {
+            return Optional.of(parsedCommands.get(parsedCommands.size() - 1).getCommand());
+        } catch (CommandLine.UnmatchedArgumentException e) {
+            String errorMessage = e.getMessage();
+            if (errorMessage != null) {
```

Comment:
shall we change this logic like (errorMessage == null) throw error else continue?